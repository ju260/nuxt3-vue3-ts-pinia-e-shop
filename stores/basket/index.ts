import { defineStore } from 'pinia'
import Product from '@/types/product';

function callApi (ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

interface State {
    items: Product[]
  }
  
  export const useBasket = defineStore('basket', {
    state: (): State => {
      return {
        items: [],
      }
    },
    actions: {
      async fetchItems () {
        try {
          await callApi(3000)
          const { data }  = await useFetch(`http://${window.location.host}/cart.json`);
          this.items = Object.values(data.value.products)
        } catch (error) {
          console.error(error)
          throw error
        }
      },
    
      async updateItem ({ itemRef, patch }) {
        try {
          await callApi(1000) // l'api doit renvoyer l'item mis a jour
          const updatedItem = this.items.filter(item => item.reference === itemRef)
          if (!updatedItem.length) return
          Object.assign(updatedItem[0], patch)
        } catch (error) {
          console.error('error during updateItem', error)
          throw error
        }
      },
      async deleteItem ({ itemRef }) {
        try {
          await callApi(1000) // le produit doit etre delete du panier cotÃ© api
          this.items = this.items.filter(item => item.reference !== itemRef)
        } catch (error) {
          console.error('error during deleteItem', error)
          throw error
        }
      }
    },
  })
  