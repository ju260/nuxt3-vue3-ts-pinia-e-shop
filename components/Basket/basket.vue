<template>
  <div class="">
    <ul class="grid gap-10">
      <li
        class=""
        v-for="(item, indexItem) in items"
        :key="`item__${indexItem}`"
      >
        <Product-item :item="item" />
      </li>
    </ul>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from 'pinia';
import { useBasket } from '@/stores/basket';

const store = useBasket();
const { items } = storeToRefs(store);
</script>
