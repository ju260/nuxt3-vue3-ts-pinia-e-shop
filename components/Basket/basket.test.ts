import Component from './basket.vue';
import products from './__fixtures__/products.js';
import { shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { describe, it, expect, beforeEach } from 'vitest';
import { useBasket } from '@/stores/basket';
import { setActivePinia, createPinia } from 'pinia';

describe('basket', () => {
  let wrapper = null;
  let store = null;

  beforeEach(() => {
    wrapper = shallowMount(Component, {
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
            initialState: {
              items: products,
            },
          }),
        ],
      },
    });

    // create a fresh Pinia instance and make it active so it's automatically picked
    // up by any useBasket() call without having to pass it to it:
    // `useBasket(pinia)`
    setActivePinia(createPinia());

    // create an instance of the data store
    store = useBasket();
  });
  describe('render', () => {
    it('should render component ', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

});
