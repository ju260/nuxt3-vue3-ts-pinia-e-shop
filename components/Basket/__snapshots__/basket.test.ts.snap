// Vitest Snapshot v1

exports[`basket > render > should render component  1`] = `
<div
  class=""
>
  <ul
    class="grid gap-10"
  >
    
    
  </ul>
</div>
`;
